src/lazy/text/matched.rs

```diff
@@ -451,3 +455,290 @@ impl MatchedSymbol {
+                let minutes_text = matched_input.slice(minutes_start, 2).as_text().unwrap();
+                let minutes = i32::from_str(minutes_text).unwrap();
+                let offset_magnitude_minutes = (hours * 60) + minutes;
+                if matched_offset.is_negative {
+                    Some(-offset_magnitude_minutes)
+                } else {
+                    Some(offset_magnitude_minutes)
+                }
+            }
+        };
+
+        let hour_text = matched_input.slice(11, 2).as_text().unwrap();
+        let hour = u32::from_str(hour_text).unwrap();
+        let minute_text = matched_input.slice(14, 2).as_text().unwrap();
+        let minute = u32::from_str(minute_text).unwrap();
+
+        let timestamp = timestamp.with_hour_and_minute(hour, minute);
+
+        if self.precision == TimestampPrecision::HourAndMinute {
+            if let Some(offset) = offset_minutes {
+                return timestamp.with_offset(offset).build();
+            } else {
+                return timestamp.build();
+            }
+        }
+
+        let second_text = matched_input.slice(17, 2).as_text().unwrap();
+        let seconds = u32::from_str(second_text).unwrap();
+        let timestamp = timestamp.with_second(seconds);
+
+        // There's guaranteed to be a 19th byte. It will either be a `.`, indicating the beginning
+        // of fractional seconds or a `+`/`-`/`Z` comprising part of the offset.
+        if matched_input.bytes()[19] != b'.' {
+            // There are no fractional seconds; we can apply the offset (if any) and return.
+            if let Some(offset) = offset_minutes {
+                return timestamp.with_offset(offset).build();
+            } else {
+                return timestamp.build();
+            }
+        }
+
+        // There are fractional seconds. We need to scan ahead to the first non-digit byte in the
+        // input following the leading `.`.
+        let fractional_start = 20;
+        let mut fractional_end = fractional_start;
+        for byte in matched_input.slice_to_end(fractional_start).bytes() {
+            if !byte.is_dec_digit() {
+                break;
+            }
+            fractional_end += 1;
+        }
+        let fractional_text = matched_input
+            .slice(fractional_start, fractional_end - fractional_start)
+            .as_text()
+            .unwrap();
+        let timestamp = match fractional_text.len() {
+            3 => {
+                let milliseconds = u32::from_str(fractional_text).unwrap();
+                timestamp.with_milliseconds(milliseconds)
+            }
+            6 => {
+                let microseconds = u32::from_str(fractional_text).unwrap();
+                timestamp.with_microseconds(microseconds)
+            }
+            9 => {
+                let nanoseconds = u32::from_str(fractional_text).unwrap();
+                timestamp.with_nanoseconds(nanoseconds)
+            }
```

Comment:
I think it would be nice to avoid creating a `Decimal` if we don't need to. I don't know how costly this would be, though. It's a small enough range that we could have a hard coded match for mapping `1-9` to powers of ten if need be.
```suggestion
            len if len <= 9 => {
                let fraction = u32::from_str(fractional_text).unwrap();
                let multiplier = pow(10, 9 - len);
                let nanoseconds = fraction * multiplier; 
                self.with_nanoseconds_and_precision(nanoseconds, len.to_u32().unwrap())
            }
```